ra-tls/libenclave-tls/enclave-tls.c

```diff
@@ -0,0 +1,262 @@
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <dlfcn.h>
+
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <unistd.h>
+
+#include <wolfssl/ssl.h>
+#include <wolfssl/certs_test.h>
+#include "wolfssl/ssl.h"
+#include "wolfssl/wolfcrypt/settings.h"
+#include "wolfssl/wolfcrypt/types.h"
+#include "wolfcrypt/test/test.h"
+#include "wolfcrypt/benchmark/benchmark.h"
+
+#include "enclave-tls.h"
+#include "wolfssl-ra-common.h"
+#include "wolfssl-ra-attester-common.h"
+#include "ra-common.h"
+
+#define DEFAULT_PORT 11111
+#define CIPHER_LIST "ECDHE-ECDSA-AES128-GCM-SHA256"
+
+// function pointer definitions
+typedef void (*wolfSSL_free_func)(WOLFSSL *);
+typedef void (*wolfSSL_CTX_set_verify_func)(WOLFSSL_CTX *, int, VerifyCallback);
+typedef int (*wolfSSL_read_func)(WOLFSSL *, void *data, int);
+typedef int (*wolfSSL_write_func)(WOLFSSL *, const void *, int);
+typedef WOLFSSL_CTX *(*wolfSSL_CTX_new_func)(WOLFSSL_METHOD *);
+typedef WOLFSSL_METHOD *(*wolfTLSv1_2_client_method_func)(void);
+typedef WOLFSSL_METHOD *(*wolfTLSv1_2_server_method_func)(void);
+typedef const unsigned char *(*wolfSSL_X509_get_der_func)(WOLFSSL_X509 *,
+							  int *);
+typedef WOLFSSL_X509 *(*wolfSSL_get_peer_certificate_func)(WOLFSSL *);
+typedef int (*wolfSSL_connect_func)(WOLFSSL *);
+typedef int (*wolfSSL_negotiate_func)(WOLFSSL *);
+typedef int (*wolfSSL_set_fd_func)(WOLFSSL *, int);
+typedef WOLFSSL *(*wolfSSL_new_func)(WOLFSSL_CTX *);
+typedef void (*wolfSSL_CTX_free_func)(WOLFSSL_CTX *);
+typedef int (*wolfSSL_Debugging_ON_func)(void);
+typedef void (*wolfSSL_Debugging_OFF_func)(void);
+typedef int (*wolfSSL_Init_func)(quote_type_t);
+typedef void (*create_key_and_x509_func)(WOLFSSL_CTX *);
+typedef int (*wolfSSL_Cleanup_func)(void);
+typedef void (*printf_identity_func)(unsigned char *, int);
+typedef int (*verify_cert_extensions_func)(uint8_t *, uint32_t);
+
+// dynamically resolved functions from libra-tls-sgx libraries
+typedef struct enclave_tls_function_table {
+	printf_identity_func printf_identity;
+	wolfSSL_free_func wolfSSL_free;
+	wolfSSL_CTX_set_verify_func wolfSSL_CTX_set_verify;
+	create_key_and_x509_func create_key_and_x509;
+	wolfSSL_read_func wolfSSL_read;
+	wolfSSL_write_func wolfSSL_write;
+	wolfSSL_CTX_new_func wolfSSL_CTX_new;
+	wolfTLSv1_2_client_method_func wolfTLSv1_2_client_method;
+	wolfSSL_X509_get_der_func wolfSSL_X509_get_der;
+	wolfSSL_get_peer_certificate_func wolfSSL_get_peer_certificate;
+	wolfSSL_connect_func wolfSSL_connect;
+	wolfSSL_negotiate_func wolfSSL_negotiate;
+	wolfSSL_set_fd_func wolfSSL_set_fd;
+	wolfSSL_new_func wolfSSL_new;
+	wolfTLSv1_2_server_method_func wolfTLSv1_2_server_method;
+	wolfSSL_CTX_free_func wolfSSL_CTX_free;
+	wolfSSL_Cleanup_func wolfSSL_Cleanup;
+	wolfSSL_Debugging_ON_func wolfSSL_Debugging_ON;
+	wolfSSL_Debugging_OFF_func wolfSSL_Debugging_OFF;
+	wolfSSL_Init_func wolfSSL_Init;
+} enclave_tls_function_table;
```

Comment:
Essentially, this is just wolfssl specific function table rather than a common function table cross all tls libraries (e.g, openssl). I mean enclave tls will support and encapsulate multiple underlaying tls libraries as the infrastructure of pure TLS protocol, so please rename this type of function table to another name, without enclave_tls prefix.

Therefore, this type is not deserved to stay the common place like enclave-tls.c. You can move them to tls_wrapper_wolfssl.c and tls_wrapper_wolfssl.h.